pkg/package-server/provider/registry.go

```diff
@@ -0,0 +1,533 @@
+		logger.Errorf("could not connect to registry service: %s", err)
+		return
+	}
+
+	func() {
+		p.cmu.Lock()
+		defer p.cmu.Unlock()
+		p.clients = append(p.clients, newRegistryClient(catsrc, conn))
+	}()
+
+	pkgList, err := p.List(catsrc.GetNamespace())
+	if err != nil {
+		logger.Warn("could not list packages to generate add events")
+		return
+	}
+	pkgs := pkgList.Items
+
+	if catsrc.GetNamespace() != p.globalNamespace {
+		// Add global packages
+		globalList, err := p.List(catsrc.GetNamespace())
+		if err != nil {
+			logger.Warn("could not list global packages to generate add events")
+			return
+		}
+		pkgs = append(pkgs, globalList.Items...)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultEventTimeout)
+	go func() {
+		defer cancel()
+		p.publishEvents(ctx, logger, pkgs, catsrc.GetNamespace(), addPkg)
+	}()
+}
+
+func (p *RegistryProvider) catalogSourceUpdated(obj interface{}, newObj interface{}) {
+	catsrc, ok := newObj.(*operatorsv1alpha1.CatalogSource)
+	if !ok {
+		logrus.Debugf("casting catalog source failed: wrong type: %#v", newObj)
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
+		"action":    "CatalogSource Updated",
+		"name":      catsrc.GetName(),
+		"namespace": catsrc.GetNamespace(),
+	})
+	logger.Info("attempting to update gRPC connection")
+
+	found := func() bool {
+		p.cmu.Lock()
+		defer p.cmu.Unlock()
+
+		for i, client := range p.clients {
+			if client.source.GetName() == catsrc.GetName() && client.source.GetNamespace() == catsrc.GetNamespace() {
+				client.conn.Close()
+				conn, err := grpc.Dial(catsrc.Status.RegistryServiceStatus.Address(), grpc.WithInsecure())
+				if err != nil {
+					logger.Errorf("could not connect to registry service: %s", err)
+				}
+				p.clients[i] = newRegistryClient(catsrc, conn)
+
+				return true
+			}
+		}
+		return false
+	}()
+
+	if !found {
+		logger.Errorf("gRPC connection not found")
+		return
+	}
+
+	pkgs, err := p.List(catsrc.GetNamespace())
+	if err != nil {
+		logger.Warn("could not list packages to generate update events")
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultEventTimeout)
+	go func() {
+		defer cancel()
+		p.publishEvents(ctx, logger, pkgs.Items, catsrc.GetNamespace(), updatePkg)
+	}()
+}
+
+func (p *RegistryProvider) catalogSourceDeleted(obj interface{}) {
+	catsrc, ok := obj.(*operatorsv1alpha1.CatalogSource)
+	if !ok {
+		logrus.Debugf("casting catalog source failed: wrong type: %#v", obj)
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
+		"action":    "CatalogSource Deleted",
+		"name":      catsrc.GetName(),
+		"namespace": catsrc.GetNamespace(),
+	})
+	logger.Debugf("attempting to remove gRPC connection")
+
+	pkgs, err := p.List(catsrc.GetNamespace())
+	if err != nil {
+		logger.Warn("could not list packages to generate events for")
+		pkgs = &v1alpha1.PackageManifestList{}
+	}
+
+	found := func() bool {
+		p.cmu.Lock()
+		defer p.cmu.Unlock()
+
+		for i, client := range p.clients {
```

Comment:
I think a map would work better here; you won't have to loop through the clients, and you can just `delete()` the key